### ISSUE TITLE
Getting multiple error message when parsing a simple templ file

### ISSUE THREAD

**Issue by pharrisee** [a-h/templ#100]
Hi, I have a very simple (copied from the examples) template:

```
package templates

templ Hello(name string) {
    <div>Hello, { name }</div>
}
```

Which causes multiple repeating popups when the file in question is in a visible VSCode tab, namely:

```
[Error - 12:13:45 PM] Request textDocument/hover failed.
  Message: no package metadata for file file:///home/user/Dev/templTest/templates/main.templ
  Code: 0 
[Error - 12:13:46 PM] Request textDocument/semanticTokens/range failed.
  Message: semantictokens are disabled
  Code: 0
[Error - 12:13:46 PM] Request textDocument/semanticTokens/full failed.
  Message: semantictokens are disabled
  Code: 0 
[Error - 12:16:41 PM] Request textDocument/semanticTokens/range failed.
  Message: semantictokens are disabled
  Code: 0 
[Error - 12:16:42 PM] Request textDocument/semanticTokens/full failed.
  Message: semantictokens are disabled
  Code: 0 
```

These popups happen every time I switch to the template in the editor, and on every change to the templ file in question.

I've installed the VSCode extension and the templ cli application.  

Can you think of anything I've missed that would cause this?

**Comment by msonawane** [comment]
9 AM UK time is fine thank you

**Comment by a-h** [comment]
https://meet.google.com/dne-xtde-ncq

**Comment by a-h** [comment]
OK, got to get on with some work. I'm out of the meet.

**Comment by msonawane** [comment]
oops it didnt notify me that you have joined i am sorry

**Comment by a-h** [comment]
No problem, we could try again, same time next week! 😁

**Comment by a-h** [comment]
In the meantime, I've updated the VS Code extension to use the latest versions of all the VS Code dependencies. Not sure if that will solve anything, it's just a regular chore to keep the JavaScript dependencies up to date.

**Comment by a-h** [comment]
I just upgraded `gopls` to v0.12.2 from v0.12.1 and got the error.

<img width="497" alt="Screenshot 2023-06-19 at 20 27 34" src="https://github.com/a-h/templ/assets/1029947/ea7a4c83-06fb-4d7a-a4b9-6fc0f503442f">

I've got a full log from the templ side: [templ.log](https://github.com/a-h/templ/files/11792609/templ.log)

And also from the gopls side: [gopls.log](https://github.com/a-h/templ/files/11792621/gopls.log)

This is a salient section of the gopls logging. This is a request/response pair from the LSP library (note the ID of 4 being used to correlate the request/response):

```
[Trace - 20:27:28.400 PM] Sending request 'textDocument/semanticTokens/range - (4)'.
Params: {"textDocument":{"uri":"file:///Users/adrian/github.com/a-h/templ/generator/test-import/template.templ"},"range":{"start":{"line":0,"character":0},"end":{"line":24,"character":1}}}

[Error - Received] 20:27:28.401 PM #4 semantictokens are disabled
```

The first thing that happens in the LSP set up is that the editor sends an init request to the LSP with its capabilities, and the server responds with what it can do.

The server is saying that it can do "semanticTokensProvider" stuff:

```json
{
  "capabilities": {
    "textDocumentSync": {
      "openClose": true,
      "change": 2,
      "save": {}
    },
    "completionProvider": {
      "triggerCharacters": [
        "."
      ]
    },
    "hoverProvider": true,
    "signatureHelpProvider": {
      "triggerCharacters": [
        "(",
        ","
      ]
    },
    "definitionProvider": true,
    "typeDefinitionProvider": true,
    "implementationProvider": true,
    "referencesProvider": true,
    "documentHighlightProvider": true,
    "documentSymbolProvider": true,
    "codeActionProvider": {
      "codeActionKinds": [
        "quickfix",
        "refactor.extract",
        "refactor.rewrite",
        "source.fixAll",
        "source.organizeImports"
      ]
    },
    "codeLensProvider": {},
    "documentLinkProvider": {},
    "workspaceSymbolProvider": true,
    "documentFormattingProvider": true,
    "renameProvider": {
      "prepareProvider": true
    },
    "foldingRangeProvider": true,
    "selectionRangeProvider": true,
    "executeCommandProvider": {
      "commands": [
        "gopls.add_dependency",
        "gopls.add_import",
        "gopls.apply_fix",
        "gopls.check_upgrades",
        "gopls.edit_go_directive",
        "gopls.fetch_vulncheck_result",
        "gopls.gc_details",
        "gopls.generate",
        "gopls.go_get_package",
        "gopls.list_imports",
        "gopls.list_known_packages",
        "gopls.mem_stats",
        "gopls.regenerate_cgo",
        "gopls.remove_dependency",
        "gopls.reset_go_mod_diagnostics",
        "gopls.run_go_work_command",
        "gopls.run_govulncheck",
        "gopls.run_tests",
        "gopls.start_debugging",
        "gopls.test",
        "gopls.tidy",
        "gopls.toggle_gc_details",
        "gopls.update_go_sum",
        "gopls.upgrade_dependency",
        "gopls.vendor",
        "gopls.workspace_stats"
      ]
    },
    "callHierarchyProvider": true,
    "semanticTokensProvider": {
      "legend": {
        "tokenTypes": [
          "namespace",
          "type",
          "class",
          "enum",
          "interface",
          "struct",
          "typeParameter",
          "parameter",
          "variable",
          "property",
          "enumMember",
          "event",
          "function",
          "method",
          "macro",
          "keyword",
          "modifier",
          "comment",
          "string",
          "number",
          "regexp",
          "operator",
          "decorator"
        ],
        "tokenModifiers": [
          "declaration",
          "definition",
          "readonly",
          "static",
          "deprecated",
          "abstract",
          "async",
          "modification",
          "documentation",
          "defaultLibrary"
        ]
      },
      "range": true,
      "full": true
    },
    "inlayHintProvider": {},
    "workspace": {
      "workspaceFolders": {
        "supported": true,
        "changeNotifications": "workspace/didChangeWorkspaceFolders"
      }
    }
  },
  "serverInfo": {
    "name": "gopls",
    "version": "{\"GoVersion\":\"go1.20.1\",\"Path\":\"golang.org/x/tools/gopls\",\"Main\":{\"Path\":\"golang.org/x/tools/gopls\",\"Version\":\"v0.12.2\",\"Sum\":\"h1:s4mznBqCHVHeuTYjxGYJhsW3Wp4FYon8YAqiJlwlpTY=\",\"Replace\":null},\"Deps\":[{\"Path\":\"github.com/BurntSushi/toml\",\"Version\":\"v1.2.1\",\"Sum\":\"h1:9F2/+DoOYIOksmaJFPw1tGFy1eDnIJXg+UHjuD8lTak=\",\"Replace\":null},{\"Path\":\"github.com/google/go-cmp\",\"Version\":\"v0.5.9\",\"Sum\":\"h1:O2Tfq5qg4qc4AmwVlvv0oLiVAGB7enBSJ2x2DqQFi38=\",\"Replace\":null},{\"Path\":\"github.com/sergi/go-diff\",\"Version\":\"v1.1.0\",\"Sum\":\"h1:we8PVUC3FE2uYfodKH/nBHMSetSfHDR6scGdBi+erh0=\",\"Replace\":null},{\"Path\":\"golang.org/x/exp\",\"Version\":\"v0.0.0-20220722155223-a9213eeb770e\",\"Sum\":\"h1:+WEEuIdZHnUeJJmEUjyYC2gfUMj69yZXw17EnHg/otA=\",\"Replace\":null},{\"Path\":\"golang.org/x/exp/typeparams\",\"Version\":\"v0.0.0-20221212164502-fae10dda9338\",\"Sum\":\"h1:2O2DON6y3XMJiQRAS1UWU+54aec2uopH3x7MAiqGW6Y=\",\"Replace\":null},{\"Path\":\"golang.org/x/mod\",\"Version\":\"v0.10.0\",\"Sum\":\"h1:lFO9qtOdlre5W1jxS3r/4szv2/6iXxScdzjoBMXNhYk=\",\"Replace\":null},{\"Path\":\"golang.org/x/sync\",\"Version\":\"v0.2.0\",\"Sum\":\"h1:PUR+T4wwASmuSTYdKjYHI5TD22Wy5ogLU5qZCOLxBrI=\",\"Replace\":null},{\"Path\":\"golang.org/x/sys\",\"Version\":\"v0.8.0\",\"Sum\":\"h1:EBmGv8NaZBZTWvrbjNoL6HVt+IVy3QDQpJs7VRIw3tU=\",\"Replace\":null},{\"Path\":\"golang.org/x/text\",\"Version\":\"v0.9.0\",\"Sum\":\"h1:2sjJmO8cDvYveuX97RDLsxlyUxLl+GHoLxBiRdHllBE=\",\"Replace\":null},{\"Path\":\"golang.org/x/tools\",\"Version\":\"v0.9.4-0.20230601215739-41a7dc76dd8f\",\"Sum\":\"h1:rZCNEwRXLwl4xnXOEYY7jTuTh3+xNnsq8IfB2idOhD0=\",\"Replace\":null},{\"Path\":\"golang.org/x/vuln\",\"Version\":\"v0.0.0-20230110180137-6ad3e3d07815\",\"Sum\":\"h1:A9kONVi4+AnuOr1dopsibH6hLi1Huy54cbeJxnq4vmU=\",\"Replace\":null},{\"Path\":\"honnef.co/go/tools\",\"Version\":\"v0.4.2\",\"Sum\":\"h1:6qXr+R5w+ktL5UkwEbPp+fEvfyoMPche6GkOpGHZcLc=\",\"Replace\":null},{\"Path\":\"mvdan.cc/gofumpt\",\"Version\":\"v0.4.0\",\"Sum\":\"h1:JVf4NN1mIpHogBj7ABpgOyZc65/UUOkKQFkoURsz4MM=\",\"Replace\":null},{\"Path\":\"mvdan.cc/xurls/v2\",\"Version\":\"v2.4.0\",\"Sum\":\"h1:tzxjVAj+wSBmDcF6zBB7/myTy3gX9xvi8Tyr28AuQgc=\",\"Replace\":null}],\"Settings\":[{\"Key\":\"-buildmode\",\"Value\":\"exe\"},{\"Key\":\"-compiler\",\"Value\":\"gc\"},{\"Key\":\"CGO_ENABLED\",\"Value\":\"1\"},{\"Key\":\"CGO_CFLAGS\",\"Value\":\"\"},{\"Key\":\"CGO_CPPFLAGS\",\"Value\":\"\"},{\"Key\":\"CGO_CXXFLAGS\",\"Value\":\"\"},{\"Key\":\"CGO_LDFLAGS\",\"Value\":\"\"},{\"Key\":\"GOARCH\",\"Value\":\"arm64\"},{\"Key\":\"GOOS\",\"Value\":\"darwin\"}],\"Version\":\"v0.12.1\"}"
  }
}
```

Later on, the templ logs clearly show that it's sending a semantic tokens request over to the LSP:

```
{"level":"info","ts":"2023-06-19T20:27:28+01:00","caller":"proxy/server.go:885","msg":"client -> server: SemanticTokensFull"}
{"level":"info","ts":"2023-06-19T20:27:28+01:00","caller":"proxy/server.go:887","msg":"client -> server: SemanticTokensFull end"}
```

And the LSP logs say that it shouldn't be!

```
[Trace - 20:27:28.402 PM] Sending request 'textDocument/semanticTokens/full - (5)'.
Params: {"textDocument":{"uri":"file:///Users/adrian/github.com/a-h/templ/generator/test-import/template.templ"}}

[Error - Received] 20:27:28.402 PM #5 semantictokens are disabled
```

So... something is likely going wrong in the negotiation, or the proxying of the information back to the LSP.

One way to deal with things is to rip out semantic token responses here:

https://github.com/a-h/templ/blob/66651f5db493f344f0ccfb1f5f5d13beba6a6fe5/cmd/templ/lspcmd/proxy/server.go#L163-L182

Or, to not pass on the request to `gopls` here, and return a default response: https://github.com/a-h/templ/blob/66651f5db493f344f0ccfb1f5f5d13beba6a6fe5/cmd/templ/lspcmd/proxy/server.go#L884-L888

But that's not getting at the real problem.

I noticed that the URI here has a `.templ` file extension, not a `.go` file extension... i.e. the filename hasn't been re-written for `gopls` like it is in the other requests:

https://github.com/a-h/templ/blob/66651f5db493f344f0ccfb1f5f5d13beba6a6fe5/cmd/templ/lspcmd/proxy/server.go#L754

So... it might be as simple as doing that.

**Comment by a-h** [comment]
Passing the Go file name didn't make any difference:

```
[Trace - 21:02:03.286 PM] Sending request 'textDocument/semanticTokens/full - (7)'.
Params: {"textDocument":{"uri":"file:///Users/adrian/github.com/a-h/templ/generator/test-import/template_templ.go"}}

[Error - Received] 21:02:03.286 PM #7 semantictokens are disabled
```

So... it's clear I've got to wipe out the request completely. I also figured out that templ is sometimes passing the templ file location on to gopls if the editor is hovering over HTML sections - it doesn't ignore the response like it should.

Easy fixes, I think. I'll look into this again, but time to stop for the day for now.

**Comment by a-h** [comment]
Hi @pharrisee, the VS Code theme did make a difference to me, the new theme seems to do something with semantic tokens, while the old one didn't.

Anyway, I've released version v0.2.304 of the templ CLI with some minor amends which I think will stop the noisy and pointless errors. No update to the VS Code extension is required, since it just uses the CLI to do its work.

I would like, at some point, the VS Code extension to check that the latest version of the templ CLI is being used.

**Comment by msonawane** [comment]
I can confirm latest binary solves the issues. Thank you @a-h

**Comment by pharrisee** [comment]
> Anyway, I've released version v0.2.304 of the templ CLI with some minor amends which I think will stop the noisy and pointless errors. No update to the VS Code extension is required, since it just uses the CLI to do its work.

Thanks for looking at this @a-h, I'll give the new version a go this morning.
